lib/adapters/adapter.ex

```diff
@@ -1,76 +1,89 @@
 defmodule Janus.Transport.WS.Adapter do
   @moduledoc """
-  This module takes part in communicating `Janus.Transport.WS` module
+  This module specifies the behaviour for adapter modules communicating `Janus.Transport.WS` module
   with lower level WebSocket client (e.g. `:websockex`).
 
-  It is responsible for sending and passing back messages, notifying about socket status change.
+  An adapter is responsible for sending and passing WebSocket frames and notifying about connection's status change (eg. disconnected).
+
+  Sending and receiving frames is supposed to be asynchronous.
+  Frames received from websocket should be forwarded to `message_receiver` process via `forward_frame/2`.
+
+  ## Creating custom adapter
+  To implement custom adapter one should use the `Janus.Transport.WS.Adapter` and implement all callbacks.
 
-  Sending messages is supposed to be synchronous while receiving is asynchronous.
-  Messages received from websocket should be forwarded to `message_receiver` process via `forward_response/2`.
 
   ## Example
   ```elixir
-  defmodule EchoAdapter do
+  # this example contains some pseudo code
+  defmodule CustomAdapter do
     use Janus.Transport.WS.Adapter
 
     @impl true
-    def connect(url, receiver, _opts) do
-      Agent.start_link(fn -> receiver end)
+    def connect(url, receiver, opts) do
+      start_websocket_connection(url, receiver, opts)
     end
 
     @impl true
-    def send(fake_socket, payload) do
-      receiver = Agent.get(fake_socket, fn receiver -> receiver end)
-
-      forward_response(receiver, payload)
-      :ok
+    def send(websocket, payload) do
+      send_frame(websocket, payload)
     end
 
     @impl true
-    def disconnect(fake_socket) do
-      receiver = Agent.get(fake_socket, fn receiver -> receiver end)
+    def disconnect(websocket) do
+      receiver = get_receiver(websocket)
+      :ok = disconnect_websocket(websocket)
```

Comment:
This example showed a few more things to change:
- `websocket` is actually a state of the adapter, could be something else than pid (e.g. a map with receiver pid and WS pid)
- as state, `websocket` should be the last argument in callbacks
- `handle_frame` is quite confusing here (only when you know the same module is passed as a callback module to library it starts to make sense)

I'd make this example more abstract but more readable I think (feel free to adjust the following suggestion)
```elixir
defmodule CustomAdapter do
  use Janus.Transport.WS.Adapter

  @impl true
  def connect(url, receiver, opts) do
    on_receive = fn msg -> forward_message(receiver, msg) end
    ws_pid = SomeLibrary.spawn_websocket_connection(url, on_receive, opts)
    %{websocket: ws_pid, receiver: receiver}
  end

  @impl true
  def send(payload, %{websocket: pid}) do
    SomeLib.send_frame(pid, payload)
   end

  @impl true
  def disconnect(state) do
    :ok = SomeLib.disconnect(state.pid)
    notify_status(state.receiver, {:disconnected, "Disconnected request"})
  end
end
```